### PR TITLE
Add MagicLink authentication mechanism to ATO login request

### PR DESCRIPTION
**Pull request by ilytvy**

### CODE REVIEW

File: lib/ravelin/authentication_mechanisms/magic_link.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+module Ravelin
+  module AuthenticationMechanisms
+    class MagicLink < RavelinObject
+      DELIVERING_METHODS = %w(email sms)
```
Comment (pedro-rodrigues-roo) [inline lib/ravelin/authentication_mechanisms/magic_link.rb:4]: Please change this to `TRANSPORTATION_MECHANISM` (how Ravelin refers to this)

File: lib/ravelin/authentication_mechanisms/magic_link.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      FAILURE_REASONS = %w(INVALID_LINK TIMEOUT INTERNAL_ERROR RATE_LIMIT BANNED_USER)
+
+      attr_accessor :transport, :success, :email, :phone_number, :failure_reason
+      attr_required :transport, :success
+
+      def failure_reason=(reason)
+        @failure_reason = reason.to_s.upcase
+      end
+
+      def validate
```
Comment (pedro-rodrigues-roo) [inline lib/ravelin/authentication_mechanisms/magic_link.rb:14]: We need to validate the presence of `phoneNumber` if `transport == "sms"` and the presence of `email` if `transport == "email"`, to make sure we comply with the API definition

File: spec/ravelin/authentication_mechanisms/magic_link_spec.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+require 'spec_helper'
+
+describe Ravelin::AuthenticationMechanisms::MagicLink do
```
Comment (pedro-rodrigues-roo) [inline spec/ravelin/authentication_mechanisms/magic_link_spec.rb:3]: Also, please add some test cases for the absence of `phoneNumber` for `transport == "sms"`, and the absence of `email` for `transport == "email"`

**Review by pedro-rodrigues-roo — CHANGES_REQUESTED** [review]

File: lib/ravelin/authentication_mechanisms/magic_link.rb (hunk 1 above)
Comment (ilytvy) [reply lib/ravelin/authentication_mechanisms/magic_link.rb]: Fixed, thanks

File: lib/ravelin/authentication_mechanisms/magic_link.rb (hunk 2 above)
Comment (ilytvy) [reply lib/ravelin/authentication_mechanisms/magic_link.rb:14]: Added, thanks

File: spec/ravelin/authentication_mechanisms/magic_link_spec.rb (hunk 3 above)
Comment (ilytvy) [reply spec/ravelin/authentication_mechanisms/magic_link_spec.rb:3]: Added, thanks

File: spec/ravelin/authentication_mechanisms/magic_link_spec.rb [hunk 4]
```diff
@@ -2,27 +2,43 @@
...
+
+  context 'fails if required transport param is not provided' do
+    subject { described_class.new(phone_number: '+447123456789', success: false, failure_reason: 'INVALID_LINK') }
+    it { expect { subject }.to raise_error(ArgumentError).with_message(/missing parameters: transport/) }
+  end
+
+  context 'fails if transport param is sms but phone_number is not provided' do
+    subject { described_class.new(transport: 'sms', email: 'joe@example.com', success: false, failure_reason: 'INVALID_LINK') }
+    it { expect { subject }.to raise_error(ArgumentError).with_message(/phone_number must be present for sms transportation mechanism/) }
+  end
```
Comment (pedro-rodrigues-roo) [inline spec/ravelin/authentication_mechanisms/magic_link_spec.rb:27]: You're still missing the test for the `email` param not being present for email transportation mechanism

Comment (ilytvy) [reply spec/ravelin/authentication_mechanisms/magic_link_spec.rb:27]: Thanks, fixed

**Review by pedro-rodrigues-roo — APPROVED** [review]

**Pull request merged**